v1/plugin/plugin.go

```diff
@@ -356,39 +427,495 @@ type preamble struct {
+				fmt.Fprintln(w, preamble)
+			})
+
+			listener, err := net.Listen("tcp", fmt.Sprintf(":%d", httpPort))
+			if err != nil {
+				log.WithFields(
+					log.Fields{
+						"port": httpPort,
+					},
+				).Fatal("Unable to get open port")
+			}
+			defer listener.Close()
+			fmt.Printf("Preamble URL: %v\n", listener.Addr().String())
+			err = http.Serve(listener, nil)
+			if err != nil {
+				log.Fatal(err)
+			}
+		}()
+		<-pluginProxy.halt
+
+	} else if libInputOutput.args() > 0 {
+		// snapteld is starting the plugin
+		// presumably with a single arg (valid json)
+		preamble, err := printPreambleAndServe(server, meta, pluginProxy, arg.ListenPort, arg.Pprof)
+		if err != nil {
+			log.Fatal(err)
+		}
+		libInputOutput.printOut(preamble)
+		go pluginProxy.HeartbeatWatch()
+		<-pluginProxy.halt
+
+	} else {
+		// no arguments provided - run and display diagnostics to the user
+		var config Config
+		if c.IsSet("config") {
+			err := json.Unmarshal([]byte(c.String("config")), &config)
+			if err != nil {
+				log.WithFields(log.Fields{
+					"error": err,
+				}).Error("unable to parse config")
+				return err
+			}
+		}
+
+		switch pluginProxy.plugin.(type) {
+		case Collector:
+			return showDiagnostics(*meta, pluginProxy, config)
+		case StreamCollector:
+			fmt.Println("Diagnostics not currently available for streaming collector plugins.")
+		case Processor:
+			fmt.Println("Diagnostics not currently available for processor plugins.")
+		case Publisher:
+			fmt.Println("Diagnostics not currently available for publisher plugins.")
+		}
+	}
+	return nil
+}
+
+func printPreambleAndServe(srv server, m *meta, p *pluginProxy, port string, isPprof bool) (string, error) {
+	l, err := net.Listen("tcp", fmt.Sprintf(":%v", port))
 	if err != nil {
-		panic("Unable to get open port")
+		return "", err
 	}
 	l.Close()
 
 	addr := fmt.Sprintf("127.0.0.1:%v", l.Addr().(*net.TCPAddr).Port)
 	lis, err := net.Listen("tcp", addr)
 	if err != nil {
-		// TODO(danielscottt): logging
-		panic(err)
+		return "", err
 	}
 	go func() {
 		err := srv.Serve(lis)
 		if err != nil {
-			panic(err)
+			log.Fatal(err)
 		}
 	}()
+	pprofAddr := "0"
+	if isPprof {
+		pprofAddr, err = startPprof()
+		if err != nil {
+			return "", err
+		}
+	}
 	resp := preamble{
 		Meta:          *m,
 		ListenAddress: addr,
 		Type:          m.Type,
-		PprofAddress:  pprofPort,
+		PprofAddress:  pprofAddr,
 		State:         0, // Hardcode success since panics on err
 	}
 	preambleJSON, err := json.Marshal(resp)
 	if err != nil {
-		panic(err)
+		return "", err
 	}
-	libInputOutput.printOut(string(preambleJSON))
-	go p.HeartbeatWatch()
-	// TODO(danielscottt): exit code
-	<-p.halt
-	return 0
+
+	return string(preambleJSON), nil
+}
+
+// GetRPCType converts a RPCType (int) to a string
+// as described in snap/control/plugin/plugin.go
+func getRPCType(i int) string {
+	if i == 0 {
+		return "NativeRPC (0)"
+	} else if i == 2 {
+		return "GRPC (2)"
+	}
+	return "Not a valid RPC Type"
```

Comment:
@kjlyon, @jcooklin - what about StreamGRPC (value 3)? See how it is declared in [snap/control/plugin/plugin.go#L79](https://github.com/intelsdi-x/snap/blob/master/control/plugin/plugin.go#L79).
 
I have a doubt how to remember to add here the next "else..if"  for some new RPC type (future speaking) and avoid getting "Not a valid RPC Type" for this new one. **What is more, in my opinion, there is no possibilities to use snap-plugin-lib-go and having NativeRPC in the same time**. 

My suggestion:
- to assign a string representation to enum in file v1/plugin/meta.go 
- please consider available RPCTypes in snap-plugin-lib-go, see [v1/plugin/meta.go#L92](https://github.com/intelsdi-x/snap-plugin-lib-go/blob/master/v1/plugin/meta.go#L92)
- it might be provided as a method for meta struct - func (m *meta) RPCType() string { [...]}
- consider using "switch...case" instead many "if..else" to keep high readability